optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,571 @@
+                    new_layer = QuantLinear(
+                        self.bits, self.group_size, in_features, out_features, True, use_cuda_fp16=self.use_cuda_fp16
+                    )
+                else:
+                    new_layer = QuantLinear(self.bits, self.group_size, in_features, out_features, True)
+                new_layer.device = device
+                setattr(module, attr, new_layer.to(device))
+        for name1, child in module.named_children():
+            self._replace_by_quant_layers(child, names, name + "." + name1 if name != "" else name1)
+
+    @torch.no_grad()
+    def quantize_model(
+        self,
+        model: nn.Module,
+        tokenizer: Any,
+        dataset: Union[List[str], str],
+        batch_size: int = 1,
+        pad_token_id: Optional[int] = None,
+    ):
+        """
+        Quantize the model using the dataset
+
+        Args:
+            model (`nn.Module`):
+                The model to quantize
+            dataset (`Union[List[str],str]`):
+                The dataset used for quantization. You can provide your own dataset in a list of string or just use the original datasets used
+                in the paper ['wikitext2','c4'].
+            tokenizer (`Any`, defaults to `None`):
+                The tokenizer to use in order to prepare the dataset
+            batch_size (`Optional[int]`, *optional*, defaults to `1`):
+                The batch size of the dataset
+            pad_token_id (`Optional[int]`, *optional*, defaults to `None`):
+                The pad token id. Needed to prepare the dataset when `batch_size` > 1.
+
+        Returns:
+            `nn.Module`: The quantized model
+        """
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed to quantize model.")
+
+        model.eval()
+
+        # For Transformer model
+        has_config = False
+        if hasattr(model, "config"):
+            has_config = True
+        if hasattr(model, "hf_device_map"):
+            # If the model has a device_map, we don't move to model. We have already dispatch the hook that will do the work
+            has_device_map = True
+        device = get_device(model)
+
+        if has_config:
+            use_cache = model.config.use_cache
+            model.config.use_cache = False
+
+        if self.model_seqlen is None:
+            self.model_seqlen = get_seqlen(model)
+
+        # Step 1: Prepare the data
+        if isinstance(dataset, str):
+            dataset, _ = get_dataset(dataset, tokenizer, seqlen=self.model_seqlen)
+        elif isinstance(dataset, list):
+            dataset = [tokenizer(data, return_tensors="pt") for data in dataset]
+        dataset = prepare_dataset(dataset, pad_token_id=pad_token_id, batch_size=batch_size)
+
+        # Step 2: get the input of the 1st block
+        layer_inputs = []
+        layer_outputs = []
+        layer_input_kwargs = []
+
+        class Catcher(nn.Module):
```

Comment:
I don't know if you copied this bit from elsewhere, but I think you could use forward pre hooks?

Reply:
Yes ofc, i will change that